### PR TITLE
Jump Search in C++

### PR DESCRIPTION
**Pull request by praveenmylavarapu**
This pull request is for issue number #20

### CODE REVIEW

File: Searching Algorithms/Jump Search Algorithm/C++/JumpSearch.cpp [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    while (right >= i - jump)
+        if (a[right--] == value)
+            return 0;
+
+    return 1;
+}
+
+int main()
+{
+    int a[100], i = 0, n, value;
```
Comment (VamsiSangam) [inline Searching Algorithms/Jump Search Algorithm/C++/JumpSearch.cpp]: Don't use fixed size arrays. Demo arrays such as `int a[] = {2, 4, 6, 8, 9, 12}` are fine.

File: Searching Algorithms/Jump Search Algorithm/C++/JumpSearch.cpp [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+Code Contributor - Praveen Mylavarapu (saipraveen.ml)
+
+===== ===== ===== */
+
+#include <iostream>
+#include <math.h>
+
+using namespace std;
+
+bool jumpSearch(int a[100], int n, int value)
```
Comment (VamsiSangam) [inline Searching Algorithms/Jump Search Algorithm/C++/JumpSearch.cpp]: `int arr[]` must be the parameter. Return type must be `int`

File: Searching Algorithms/Jump Search Algorithm/C++/JumpSearch.cpp [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    for (i = 0; i + jump < n && a[i] < value; i += jump);
+
+    if (a[i] >= value)
+        right = i;
+
+    while (right >= i - jump)
+        if (a[right--] == value)
+            return 0;
+
+    return 1;
```
Comment (VamsiSangam) [inline Searching Algorithms/Jump Search Algorithm/C++/JumpSearch.cpp]: Return `-1` here.

File: Searching Algorithms/Jump Search Algorithm/C++/JumpSearch.cpp [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+    while (i < n)
+        cin >> a[i++];
+
+    cout << "Enter the value to be searched: ";
+    cin >> value;
+
+    if (jumpSearch(a, n, value))
+        cout << "Not ";
+    cout << "Found!" << endl;
```
Comment (VamsiSangam) [inline Searching Algorithms/Jump Search Algorithm/C++/JumpSearch.cpp]: You can just print the return value of `jumpSearch()`. Make multiple calls on same function to show different cases, like the other implementations.

**Review by VamsiSangam — CHANGES_REQUESTED** [review]
Code seems to work fine, but you implementation must resemble the other existing implementations as much as possible, eg. in terms of return values.

Try to have a look at the other implementations and keep your code familiar to it.

Also, you seem to have written code with the aim of reducing number of lines. Short code is often not easy to read. As this code will be displayed on a tutorial website, your code must be easily understandable. It is okay if you lengthen your code with full proper loops, comments or some extra code, it will only improve understand-ability.

Also, no need for the input/output test case files here, as we will show some cases in the main method itself.

**Comment by praveenmylavarapu** [comment]
I made the requested changes.
Should my implementation resemble the other existing implementations as much as possible in terms of logic also? Mine is actually simple and it does Linear Search Backwards as described in the image of the article.
And should I use braces for single line loops and conditionals?

File: Searching Algorithms/Jump Search Algorithm/C++/JumpSearch.cpp [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+#include <math.h>
+
+using namespace std;
+
+int jumpSearch(int arr[], int n, int value)
+{
+    int i, right, left, jump = (int) sqrt(n);
+
+    // Keep jumping until value is between any two elements of the array or when i goes out of bound
+    for (i = 0; i + jump < n && arr[i] < value; i += jump);
```
Comment (VamsiSangam) [inline Searching Algorithms/Jump Search Algorithm/C++/JumpSearch.cpp]: Insert the comment inside this loop, and break the long comment into multiple lines. Eg. -

``` c
for (...) {
    // Keep jumping until value is between any two
    // elements of the array or when i goes out of bound
}
```

File: Searching Algorithms/Jump Search Algorithm/C++/JumpSearch.cpp [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+    if (value <= arr[i])    // right is i if search value is less than any jumped element in the array. If not, right is last element
```
Comment (VamsiSangam) [inline Searching Algorithms/Jump Search Algorithm/C++/JumpSearch.cpp]: Here too, insert the comment inside the `if`. Eg. -

``` c
if (value <= arr[i]) {
    // right is i if search value is less than
    // any jumped element in the array.
    right = i;
}
```

File: Searching Algorithms/Jump Search Algorithm/C++/JumpSearch.cpp [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+        right = i;
+    else
+        right = n;
+
+    left = right - jump;
```
Comment (VamsiSangam) [inline Searching Algorithms/Jump Search Algorithm/C++/JumpSearch.cpp]: Use `left = max(right - jump, 0)` instead. Also add the comment `// ensuring left doesn't go negative` next to it.

File: Searching Algorithms/Jump Search Algorithm/C++/JumpSearch.cpp [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    while (right >= left)
+        if (arr[right--] == value)
+            return right + 1;
+
+    return -1;
+}
+
+int main()
+{
+    int arr[] = {1, 3, 5, 7, 9, 11, 13, 15}, n = sizeof(arr)/sizeof(int);
```
Comment (VamsiSangam) [inline Searching Algorithms/Jump Search Algorithm/C++/JumpSearch.cpp]: Write the declaration of `n` in a new line.

**Review by VamsiSangam — CHANGES_REQUESTED** [review]
Minor cosmetic changes requested.

File: Searching Algorithms/Jump Search Algorithm/C++/JumpSearch.cpp [hunk 9]
```diff
@@ -18,18 +19,21 @@ int jumpSearch(int arr[], int n, int value)
...
+    }
 
-    if (value <= arr[i])    // right is i if search value is less than any jumped element in the array. If not, right is last element
+    if (value <= arr[i]) {
+        // right is i if search value is less than
+        // any jumped element in the array. If not, right is last element
         right = i;
-    else
+    }
+    else {
```
Comment (VamsiSangam) [inline Searching Algorithms/Jump Search Algorithm/C++/JumpSearch.cpp]: `// If not, right is last element` in the previous comment will come under this else, not in the `if` clause. Also please correct the brackets. The code should be -

``` c
    if (value <= arr[i]) {
        // right is i if search value is less than
        // any jumped element in the array.
        right = i;
    } else {
        // If not, right is last element
        right = n;
    }
```

File: Searching Algorithms/Jump Search Algorithm/C++/JumpSearch.cpp [hunk 10]
```diff
@@ -18,18 +19,21 @@ int jumpSearch(int arr[], int n, int value)
...
         right = n;
+    }
 
-    left = right - jump;
-
-    if (left < 0)   // ensuring left dosen't go negaitive
-        left = 0;
+    left = max(right - jump, 0);    // ensuring left dosen't go negaitive
```
Comment (VamsiSangam) [inline Searching Algorithms/Jump Search Algorithm/C++/JumpSearch.cpp]: Correct spellings :stuck_out_tongue_closed_eyes: 

> ensuring left **doesn't** go **negative**

File: Searching Algorithms/Jump Search Algorithm/C++/JumpSearch.cpp [hunk 11]
```diff
@@ -41,7 +45,8 @@ int jumpSearch(int arr[], int n, int value)
 
 int main()
 {
-    int arr[] = {1, 3, 5, 7, 9, 11, 13, 15}, n = sizeof(arr)/sizeof(int);
+    int arr[] = {1, 3, 5, 7, 9, 11, 13, 15};
+    int n = sizeof(arr)/sizeof(int);
```
Comment (VamsiSangam) [inline Searching Algorithms/Jump Search Algorithm/C++/JumpSearch.cpp]: Add space on either side of `/` operator.

**Review by VamsiSangam — CHANGES_REQUESTED** [review]

**Comment by praveenmylavarapu** [comment]
Sorry for my silly mistakes and thank you for your patience @VamsiSangam

File: Searching Algorithms/Jump Search Algorithm/C++/JumpSearch.cpp [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+Jump Search Algorithm
+http://theoryofprogramming.com/2016/11/10/jump-search-algorithm/
+GitHub - https://github.com/VamsiSangam/theoryofprogramming
+Code Contributor - Praveen Mylavarapu (saipraveen.ml)
+
+===== ===== ===== */
+
+#include <iostream>
+#include <algorithm>
+#include <math.h>
```
Comment (VamsiSangam) [inline Searching Algorithms/Jump Search Algorithm/C++/JumpSearch.cpp]: Change this to `cmath`

File: Searching Algorithms/Jump Search Algorithm/C++/JumpSearch.cpp [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+        right = i;
+    } else {
+        // If not, right is last element
+        right = n;
+    }
+
+    left = max(right - jump, 0);    // ensuring left doesn't go negative
+
+    // Linear search backwards
+    while (right >= left)
```
Comment (VamsiSangam) [inline Searching Algorithms/Jump Search Algorithm/C++/JumpSearch.cpp]: Add `{ }` here

File: Searching Algorithms/Jump Search Algorithm/C++/JumpSearch.cpp [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+        if (arr[right--] == value)
```
Comment (VamsiSangam) [inline Searching Algorithms/Jump Search Algorithm/C++/JumpSearch.cpp]: Add `{ }` here

**Review by VamsiSangam — CHANGES_REQUESTED** [review]
Minor cosmetic changes requested.

**Review by VamsiSangam — APPROVED** [review]
LGTM :+1:

**Comment by VamsiSangam** [comment]
Thanks for your contribution! You should be able to see your code live on the [corresponding post](http://theoryofprogramming.com/2016/11/10/jump-search-algorithm/). :smile:

**Pull request merged**